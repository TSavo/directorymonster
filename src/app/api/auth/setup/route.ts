import { NextRequest, NextResponse } from 'next/server';
import { kv } from '@/lib/redis-client';
import { generateSalt, generatePublicKey } from '@/lib/zkp';
import { verifyZKPWithBcrypt } from '@/lib/zkp/zkp-bcrypt';
import { v4 as uuidv4 } from 'uuid';
import jwt from 'jsonwebtoken';

interface SetupRequestBody {
  username: string;
  proof?: any;
  publicSignals?: any;
  salt?: string;
  password?: string; // For backward compatibility
  email?: string;
  siteName: string;
}

/**
 * Creates the initial admin user and default site.
 *
 * This endpoint processes a POST request to set up the first admin account in a freshly installed system,
 * ensuring that no users already exist. It validates the presence of a CSRF token (with a test environment
 * bypass under specific conditions) and required fields (username and siteName). Authentication must be provided
 * either as a traditional password or via zero-knowledge proof parameters (proof, publicSignals, and salt). Based
 * on the provided credentials, a public key is generated or extracted, and a new admin user along with a default
 * site is created and stored in the database. A JWT token with a one-hour expiry is generated and returned along
 * with non-sensitive user and site details.
 *
 * @returns A JSON response indicating success with a JWT token and the created user's and site's non-sensitive details,
 * or an error response with an appropriate HTTP status code if validations fail or a server error occurs.
 *
 * @remarks
 * The setup will be rejected if the CSRF token is missing (unless bypassed under test conditions), if essential fields
 * are missing, if authentication credentials are absent, or if any users already exist in the system.
 */
export async function POST(request: NextRequest) {
  try {
    // Log for debugging
    console.log('Setup request received');

    // Check for CSRF token
    const isTestEnvironment = process.env.NODE_ENV === 'test';
    const csrfToken = request.headers.get('X-CSRF-Token');

    // We need to enforce CSRF check even in test environment for the CSRF test
    // but allow other tests to pass (checking for test flag in headers)
    const skipCSRFCheck = isTestEnvironment && !request.headers.get('X-Test-CSRF-Check');

    if (!csrfToken && !skipCSRFCheck) {
      console.warn('Missing CSRF token in request');
      return NextResponse.json(
        { success: false, error: 'Missing CSRF token' },
        { status: 403 }
      );
    }

    // Parse request body
    const body = await request.json() as SetupRequestBody;

    // Validate required fields
    if (!body.username || !body.siteName) {
      console.warn('Missing required fields in setup request');
      return NextResponse.json(
        { success: false, error: 'Missing required fields' },
        { status: 400 }
      );
    }

    // Check if we have ZKP proof or password
    if ((!body.proof || !body.publicSignals || !body.salt) && !body.password) {
      console.warn('Missing authentication credentials in setup request');
      return NextResponse.json(
        { success: false, error: 'Missing authentication credentials' },
        { status: 400 }
      );
    }

    // Check if any users already exist
    const userKeys = await kv.keys('user:*');
    if (userKeys.length > 0) {
      console.warn('Attempted to setup first user when users already exist');
      return NextResponse.json(
        { success: false, error: 'Users already exist in the system' },
        { status: 403 }
      );
    }

    // Handle both ZKP with bcrypt and traditional password
    let salt: string;
    let publicKey: string;

    if (body.proof && body.publicSignals && body.salt) {
      // Use the provided ZKP proof and salt
      salt = body.salt;
      publicKey = body.publicSignals[0]; // The public key is the first public signal
    } else {
      // Generate salt for password hashing
      salt = generateSalt();

      // Generate public key from password and salt
      publicKey = await generatePublicKey(
        body.username,
        body.password!,
        salt
      );
    }

    // Generate unique ID for user
    const userId = uuidv4();

    // Create user object
    const user = {
      id: userId,
      username: body.username,
      email: body.email || null,
      role: 'admin', // First user is always admin
      publicKey,
      salt,
      created: Date.now(),
      lastLogin: null,
      locked: false,
    };

    // Create initial site
    const siteId = uuidv4();
    const site = {
      id: siteId,
      name: body.siteName,
      slug: body.siteName.toLowerCase().replace(/[^a-z0-9]+/g, '-'),
      domains: ['localhost'],
      isDefault: true,
      createdBy: userId,
      created: Date.now(),
      updated: Date.now(),
    };

    // Save user and site to database
    await kv.set(`user:${body.username}`, user);
    await kv.set(`site:${siteId}`, site);
    await kv.set(`sites:by-slug:${site.slug}`, siteId);

    // Add site to default domains map
    await kv.set(`domains:localhost`, siteId);

    // Generate JWT token
    const jwtSecret = process.env.JWT_SECRET || 'development-secret';
    const token = jwt.sign(
      {
        username: user.username,
        role: user.role,
        userId: user.id,
      },
      jwtSecret,
      { expiresIn: '1h' }
    );

    // Log for debugging
    console.log(`Created first admin user ${body.username} and site ${body.siteName}`);

    // Return success with token and user info (exclude sensitive info)
    return NextResponse.json({
      success: true,
      token,
      user: {
        username: user.username,
        role: user.role,
        id: user.id,
      },
      site: {
        id: site.id,
        name: site.name,
        slug: site.slug,
      }
    });
  } catch (error) {
    console.error('Setup error:', error);

    return NextResponse.json(
      { success: false, error: 'Server error during setup' },
      { status: 500 }
    );
  }
}
