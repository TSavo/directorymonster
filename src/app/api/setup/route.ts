import { NextRequest, NextResponse } from 'next/server';
import { TenantService } from '@/lib/tenant';
import { withRedis } from '@/middleware/withRedis';

/**
 * GET handler for checking if any tenants exist
 * @param request The incoming request
 * @returns JSON response with setup status
 */
async function GET(request: NextRequest) {
  try {
    // Check if any tenants exist
    const tenantsExist = await TenantService.tenantsExist();
    
    // Return setup status
    return NextResponse.json({
      needsSetup: !tenantsExist,
      setupComplete: tenantsExist
    });
  } catch (error) {
    console.error('Error checking setup status:', error);
    return NextResponse.json(
      { error: 'Internal Server Error' },
      { status: 500 }
    );
  }
}

/**
 * POST handler for initial setup with first tenant
 * @param request The incoming request with setup data
 * @returns JSON response with setup result
 */
async function POST(request: NextRequest) {
  try {
    // Check if tenants already exist
    const tenantsExist = await TenantService.tenantsExist();
    
    if (tenantsExist) {
      return NextResponse.json(
        { error: 'Setup already completed' },
        { status: 400 }
      );
    }
    
    // Parse request body
    const setupData = await request.json();
    
    // Create first tenant
    const tenant = await TenantService.createTenant({
      slug: setupData.tenant.slug,
      name: setupData.tenant.name,
      hostnames: setupData.tenant.hostnames || ['localhost'],
      primaryHostname: setupData.tenant.primaryHostname || 'localhost',
      theme: setupData.tenant.theme || 'default',
      settings: setupData.tenant.settings || {},
      active: true,
    });
    
    // Return success with tenant data
    return NextResponse.json({
      success: true,
      message: 'Initial setup completed successfully',
      tenant,
    }, { status: 201 });
  } catch (error) {
    console.error('Error during initial setup:', error);
    return NextResponse.json(
      { error: 'Setup failed', message: error instanceof Error ? error.message : String(error) },
      { status: 500 }
    );
  }
}

// Apply Redis middleware to both handlers
const handlers = { GET, POST };

export { handlers as GET, handlers as POST };