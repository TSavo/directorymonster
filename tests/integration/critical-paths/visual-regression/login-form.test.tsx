/**
 * @jest-environment jsdom
 */

import React from 'react';
import { fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { renderForSnapshot, snapshotTest } from './setup';

// Mock the login form component
const LoginForm = () => {
  const [username, setUsername] = React.useState('');
  const [password, setPassword] = React.useState('');
  const [errors, setErrors] = React.useState<{ username?: string; password?: string }>({});
  const [isSubmitting, setIsSubmitting] = React.useState(false);
  const [isSuccess, setIsSuccess] = React.useState(false);

  const validateForm = () => {
    const newErrors: { username?: string; password?: string } = {};

    if (!username.trim()) {
      newErrors.username = 'Username is required';
    }

    if (!password) {
      newErrors.password = 'Password is required';
    } else if (password.length < 8) {
      newErrors.password = 'Password must be at least 8 characters';
    }

    setErrors(newErrors);
    return Object.keys(newErrors).length === 0;
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    if (validateForm()) {
      setIsSubmitting(true);

      // Simulate API call
      setTimeout(() => {
        setIsSubmitting(false);
        setIsSuccess(true);
      }, 500);
    }
  };

  return (
    <div className="login-form-container">
      {isSuccess ? (
        <div className="success-message" data-testid="success-message">
          <h2>Login successful!</h2>
          <p>You are now logged in.</p>
        </div>
      ) : (
        <form onSubmit={handleSubmit} className="login-form" data-testid="login-form">
          <h2>Log In</h2>

          <div className="form-group">
            <label htmlFor="username">Username</label>
            <input
              type="text"
              id="username"
              className={errors.username ? 'input-error' : ''}
              value={username}
              onChange={e => setUsername(e.target.value)}
              aria-invalid={!!errors.username}
              aria-describedby={errors.username ? 'username-error' : undefined}
              data-testid="username-input"
            />
            {errors.username && (
              <div id="username-error" className="error-message" role="alert" data-testid="username-error">
                {errors.username}
              </div>
            )}
          </div>

          <div className="form-group">
            <label htmlFor="password">Password</label>
            <input
              type="password"
              id="password"
              className={errors.password ? 'input-error' : ''}
              value={password}
              onChange={e => setPassword(e.target.value)}
              aria-invalid={!!errors.password}
              aria-describedby={errors.password ? 'password-error' : undefined}
              data-testid="password-input"
            />
            {errors.password && (
              <div id="password-error" className="error-message" role="alert" data-testid="password-error">
                {errors.password}
              </div>
            )}
          </div>

          <button
            type="submit"
            className="login-button"
            disabled={isSubmitting}
            data-testid="login-button"
          >
            {isSubmitting ? 'Logging in...' : 'Log In'}
          </button>
        </form>
      )}
    </div>
  );
};

describe('Login Form Visual Regression', () => {
  it('renders the login form correctly', () => {
    const container = renderForSnapshot(<LoginForm />);
    snapshotTest(container, 'login-form-default');
  });

  it('renders validation errors correctly', () => {
    const container = renderForSnapshot(<LoginForm />);

    // Submit the form without entering any data
    const loginButton = container.querySelector('[data-testid="login-button"]');
    if (loginButton) {
      fireEvent.click(loginButton);
    }

    snapshotTest(container, 'login-form-with-validation-errors');
  });

  it('renders the form with entered data correctly', () => {
    const container = renderForSnapshot(<LoginForm />);

    // Enter valid data
    const usernameInput = container.querySelector('[data-testid="username-input"]');
    const passwordInput = container.querySelector('[data-testid="password-input"]');

    if (usernameInput && passwordInput) {
      fireEvent.change(usernameInput, { target: { value: 'testuser' } });
      fireEvent.change(passwordInput, { target: { value: 'password123' } });
    }

    snapshotTest(container, 'login-form-with-data');
  });
});
