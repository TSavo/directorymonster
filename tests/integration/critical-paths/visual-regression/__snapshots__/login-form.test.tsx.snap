// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form Visual Regression renders the form with entered data correctly: login-form-with-data 1`] = `"<div class="login-form-container"><form class="login-form" data-testid="login-form"><h2>Log In</h2><div class="form-group"><label for="username">Username</label><input type="text" id="username" class="" aria-invalid="false" data-testid="username-input" value="testuser"></div><div class="form-group"><label for="password">Password</label><input type="password" id="password" class="" aria-invalid="false" data-testid="password-input" value="password123"></div><button type="submit" class="login-button" data-testid="login-button">Log In</button></form></div>"`;

exports[`Login Form Visual Regression renders the login form correctly: login-form-default 1`] = `"<div class="login-form-container"><form class="login-form" data-testid="login-form"><h2>Log In</h2><div class="form-group"><label for="username">Username</label><input type="text" id="username" class="" aria-invalid="false" data-testid="username-input" value=""></div><div class="form-group"><label for="password">Password</label><input type="password" id="password" class="" aria-invalid="false" data-testid="password-input" value=""></div><button type="submit" class="login-button" data-testid="login-button">Log In</button></form></div>"`;

exports[`Login Form Visual Regression renders validation errors correctly: login-form-with-validation-errors 1`] = `"<div class="login-form-container"><form class="login-form" data-testid="login-form"><h2>Log In</h2><div class="form-group"><label for="username">Username</label><input type="text" id="username" class="input-error" aria-invalid="true" data-testid="username-input" value="" aria-describedby="username-error"><div id="username-error" class="error-message" role="alert" data-testid="username-error">Username is required</div></div><div class="form-group"><label for="password">Password</label><input type="password" id="password" class="input-error" aria-invalid="true" data-testid="password-input" value="" aria-describedby="password-error"><div id="password-error" class="error-message" role="alert" data-testid="password-error">Password is required</div></div><button type="submit" class="login-button" data-testid="login-button">Log In</button></form></div>"`;
